### PR TITLE
feat(chain)!: Add `time_of_sync` to `SyncRequest` and `FullScanRequest` WIP

### PR DESCRIPTION
**Pull request by LagginTimes**
Implements #1550.

### Description

Adds `time_of_sync` as a member of `SyncRequest` and `FullSyncRequest` so we can set the `last_seen` for unconfirmed transactions during the sync/scan process in `bdk_electrum`.

### Notes to the reviewers

This PR is a WIP because setting the `last_seen` for unconfirmed transactions during the sync/scan process is yet to be implemented for `bdk_esplora`. I pushed my WIP towards setting the `last_seen` for `bdk_esplora` but I'm unsure if it is correct, as I did not have enough time to confirm and write a test for it before my travel. Please feel free to drop that commit if the concept is incorrect.

### Changelog notice

* Removed `update_last_seen_unconfirmed()` from `TxGraph`.
* Added `time_of_sync` as a member of `SyncRequest`.
* Added `time_of_sync` as a member of `FullScanRequest`.

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

### CODE REVIEW

File: crates/chain/src/spk_client.rs [hunk 1]
```diff
@@ -193,6 +193,12 @@ impl<I> SyncRequestBuilder<I> {
         self
     }
 
+    /// Set the `time_of_sync` for unconfirmed transactions.
+    pub fn time_of_sync(mut self, time_of_sync: u64) -> Self {
```
Comment (notmandatory) [inline crates/chain/src/spk_client.rs:197]: Is this value required ? If so should it be a parameter of `SyncRequest::builder()`?  Otherwise if someone forgets to set it it will always be 0. (And same for FullScanRequest::builder()).  

Or if it's not required as you have it is fine.

Comment (LagginTimes) [reply crates/chain/src/spk_client.rs:197]: At first I thought to make it user-configurable, but I think you're right, it makes significantly more sense to just have the `time_of_sync` set by default.

But we would run into the `no_std` issue we we were to set it from `SyncRequest`/`FullScanRequest`.

File: crates/chain/src/spk_client.rs [hunk 2]
```diff
@@ -239,6 +239,7 @@ pub struct SyncRequest<I = ()> {
     txids_consumed: usize,
     outpoints: VecDeque<OutPoint>,
     outpoints_consumed: usize,
+    time_of_sync: u64,
```
Comment (evanlinjin) [inline crates/chain/src/spk_client.rs]:
```suggestion
    time_of_sync: Option<u64>,
```

I think it's important that the caller can opt out of setting a seen_at for unconfirmed transactions.

Additionally, I don't think having a default of 0 is a good idea... However, I need to think more on this.

File: crates/chain/src/tx_graph.rs [hunk 3]
```diff
@@ -565,76 +565,13 @@ impl<A: Clone + Ord> TxGraph<A> {
...
-                },
-            )
-            .collect();
-
-        for txid in unanchored_txs {
-            changeset.merge(self.insert_seen_at(txid, seen_at));
-        }
-        changeset
-    }
-
```
Comment (evanlinjin) [inline crates/chain/src/tx_graph.rs:637]: I think removing this method should really be a separate commit. However, I'm wondering if there is an argument to keep it.

File: crates/chain/src/spk_client.rs (hunk 1 above)
Comment (notmandatory) [reply crates/chain/src/spk_client.rs:197]: Now that you have internal `time_of_sync` field as an `Option<u64>` it makes sense to leave this function as-is.

File: crates/chain/src/tx_graph.rs (hunk 3 above)
Comment (notmandatory) [reply crates/chain/src/tx_graph.rs:637]: My 2 sats is if both wallet and non-wallet based app will use `SyncRequest` and `FullScanRequest` then this shouldn't be needed.
